doc/sphinx/proof-engine/tactics.rst

```diff
@@ -898,211 +1109,156 @@ Applying theorems
-   type of :token:`ident` against a non-dependent premise of the type
-   of :token:`term`, trying them from right to left. If it succeeds, the
-   statement of hypothesis :token:`ident` is replaced by the conclusion of
-   the type of :token:`term`. The tactic also returns as many subgoals as the
-   number of other non-dependent premises in the type of :token:`term` and of
-   the non-dependent premises of the type of :token:`ident`. If the conclusion
-   of the type of :token:`term` does not match the goal *and* the conclusion
-   is an inductive type isomorphic to a tuple type, then
-   the tuple is (recursively) decomposed and the first component of the tuple
-   of which a non-dependent premise matches the conclusion of the type of
-   :token:`ident`. Tuples are decomposed in a width-first left-to-right order
-   (for instance if the type of :g:`H1` is :g:`A <-> B` and the type of
-   :g:`H2` is :g:`A` then :g:`apply H1 in H2` transforms the type of :g:`H2`
-   into :g:`B`). The tactic :tacn:`apply` relies on first-order pattern matching
-   with dependent types.
-
-   .. exn:: Statement without assumptions.
-
-      This happens if the type of :token:`term` has no non-dependent premise.
-
-   .. exn:: Unable to apply.
-
-      This happens if the conclusion of :token:`ident` does not match any of
-      the non-dependent premises of the type of :token:`term`.
-
-   .. tacv:: apply {+, @term} in {+, @ident}
-
-      This applies each :token:`term` in sequence in each hypothesis :token:`ident`.
-
-   .. tacv:: apply {+, @term with @bindings} in {+, @ident}
-
-      This does the same but uses the bindings to instantiate
-      parameters of :token:`term` (see :ref:`bindings`).
+.. _managingthelocalcontext:
 
-   .. tacv:: eapply {+, @term {? with @bindings } } in {+, @ident}
+Managing the local context
+------------------------------
 
-      This works as :tacn:`apply … in` but turns unresolved bindings into
-      existential variables, if any, instead of failing.
+.. tacn:: intro {? @ident } {? @where }
 
-   .. tacv:: apply {+, @term {? with @bindings } } in {+, @ident {? as @simple_intropattern}}
-      :name: apply … in … as
+   Applies the :tacn:`hnf` tactic until it finds an item that can be
+   introduced in the context by removing certain constructs in the goal.
+   If no item is found, the tactic fails.  The constructs have these forms:
+   (See examples :ref:`here <intro_examples>`.)
 
-      This works as :tacn:`apply … in` but applying an associated
-      :token:`simple_intropattern` to each hypothesis :token:`ident`
-      that comes with such clause.
+   `forall x : T, …`
+     `x` is a :term:`dependent premise`.  Removes `forall x`
+     from the goal and adds `x : T` to the context.
 
-   .. tacv:: simple apply @term in {+, @ident}
+   `A -> B`
+     `A` is a :term:`non-dependent premise`.  Removes `A ->` from
+     the goal and adds `H : A` to the context.  If `H` already exists, Coq uses
+     a fresh name.
```

Comment:
This is the only mention that Coq generates a fresh name in case the one it wanted to use already exists in the context, but this is actually true in all three cases.